### PR TITLE
GUVNOR-2603 : Guided Decision Table: V&V: Adding BRL Condition column throws IndexOutOfBounds in

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/GUVNOR-2648

### CODE REVIEW

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-api/src/main/java/org/drools/workbench/screens/guided/dtable/service/GuidedDecisionTableEditorService.java [hunk 1]
```diff
@@ -44,7 +44,7 @@
         SupportsCopy,
         SupportsRename {
 
-    public static String DTABLE_VERIFICATION_ENABLED = "org.kie.verification.disable-dtable-realtime-verification";
+    public static String DTABLE_VERIFICATION_DISABLED = "org.kie.verification.disable-dtable-realtime-verification";
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-api/src/main/java/org/drools/workbench/screens/guided/dtable/service/GuidedDecisionTableEditorService.java]: Naming was wrong here. Fixed it on the side.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/PatternInspector.java [hunk 2]
```diff
@@ -131,6 +120,13 @@ public Conflict hasConflicts() {
     }
 
     public ActionsInspectorMultiMap getActionsInspector() {
+        final ActionsInspectorMultiMap actionsInspector = new ActionsInspectorMultiMap();
+
+        for ( final FieldInspector fieldInspector : inspectorList ) {
+            actionsInspector.addAllValues( fieldInspector.getObjectField(),
+                                           fieldInspector.getActionInspectorList() );
+        }
+
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/PatternInspector.java]: Part of the problem. Even when changes to the rule happened, this blocked them from showing up in the inspections.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/RuleInspectorDumper.java [hunk 3]
```diff
@@ -68,6 +74,28 @@ private void dumpConditions() {
...
+            dump.append( "Action{\n" );
+            if ( object instanceof HumanReadable ) {
+                dump.append( (( HumanReadable ) object).toHumanReadableString() );
+            } else {
+                dump.append( object.toString() );
+            }
+            dump.append( "}\n" );
+        }
+    }
+
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/RuleInspectorDumper.java]: Just my helper class for situations where GWT debugger does not fire up ( about 7 times out of 10 ) or just when I need a more readable version of the rule inspector.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/controller/AfterColumnDeleted.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+package org.drools.workbench.screens.guided.dtable.client.widget.analysis.controller;
+
+import com.google.gwt.event.shared.EventHandler;
+import com.google.gwt.event.shared.GwtEvent;
+
+/**
+ * An event to notify a column was deleted
+ */
+public class AfterColumnDeleted
+        extends GwtEvent<AfterColumnDeleted.Handler> {
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/controller/AfterColumnDeleted.java]: Moving this closer to verifier codes. Used to be in kie-wb-common, but that event is tied to how the grid works that dtables used to be based on.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/controller/AfterColumnInserted.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.drools.workbench.screens.guided.dtable.client.widget.analysis.controller;
+
+import com.google.gwt.event.shared.EventHandler;
+import com.google.gwt.event.shared.GwtEvent;
+import org.drools.workbench.models.guided.dtable.shared.model.BaseColumn;
+
+public class AfterColumnInserted
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/controller/AfterColumnInserted.java]: Same situation that we had with `AfterColumnDeleted`

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/index/CellBuilder.java [hunk 6]
```diff
@@ -61,14 +61,23 @@ public void build( final Rule rule,
...
+                                                                  columnIndex );
+            } else {
+                new FieldConditionsBuilder( index,
+                                            model,
+                                            rule,
+                                            row,
+                                            utils,
+                                            getPattern( rule ) ).buildCondition( ( ConditionCol52 ) baseColumn,
+                                                                                 columnIndex );
+            }
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/index/CellBuilder.java]: This caused the NPE. There is not pattern to be found for the new BRL column, still I tried to build the `FieldCondition` that for fields of a pattern.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/DecisionTableAnalyzerUpdateTest.java [hunk 7]
```diff
@@ -165,17 +165,44 @@ public void testAddColumn() throws Exception {
...
+                               createBRLConditionColumn(),
+                               true,
+                               true,
+                               true );
+
+        assertDoesNotContain( "RedundantRows", analyzerProvider.getAnalysisReport() );
+
+    }
+
+    @Test
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/DecisionTableAnalyzerUpdateTest.java]: The test that proves the fix works.

**Comment by Rikkola** [comment]
Fixing the conflicts.

**Comment by Rikkola** [comment]
@karreiro if you could please review.

**Comment by karreiro** [comment]
@Rikkola +1 :-) Thanks

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/RuleInspectorCache.java [hunk 8]
```diff
@@ -65,8 +65,10 @@ public RuleInspectorCache( final ColumnUtilities utils,
 
     public void newColumn( final int columnIndex ) {
 
+        final List<BaseColumn> expandedColumns = model.getExpandedColumns();
+        final BaseColumn baseColumn1 = expandedColumns.get( columnIndex );
```
Comment (csadilek) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/RuleInspectorCache.java]: I am picky, but why is this variable called baseColumn*1*?

Comment (Rikkola) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/RuleInspectorCache.java]: Nothing wrong with being picky.

Looks like at some point I had two base columns. I often use the variables just for easier debugging, I prefer to have them directly in the method/constructor call like they were here before. Going to refactor back to that.

**Comment by Rikkola** [comment]
I left few not needed variables in, some even affected performance. Cleaned them and rebased with master.

**Pull request merged**